src/utils/test/utils.cpp

```diff
@@ -115,6 +115,52 @@ TEST(core, check_c_string_empty)
     }
 }
 
+TEST(core, check_c_string_equal)
+{
+    struct test_case
+    {
+        const char *lhs;
+        const char *rhs;
+        bool is_equal;
+    } tests[] = {
+        {nullptr, nullptr, true}, {nullptr, "", false},  {nullptr, "a", false},
+        {nullptr, "abc", false},  {"", nullptr, false},  {"a", nullptr, false},
+        {"abc", nullptr, false},  {"", "", true},        {"", "a", false},
+        {"", "abc", false},       {"a", "", false},      {"abc", "", false},
+        {"a", "a", true},         {"a", "A", false},     {"A", "A", true},
+        {"abc", "abc", true},     {"aBc", "abc", false}, {"abc", "ABC", false},
+        {"a", "abc", false},      {"A", "abc", false},   {"abc", "a", false},
+        {"Abc", "a", false},
+    };
+
+    for (const auto &test : tests) {
+        EXPECT_EQ(test.is_equal, dsn::utils::equals(test.lhs, test.rhs));
```

Comment:
This style of unit test is wildly used, but it has a shortcoming, when some case failed, it's hard to find out which one failed, right?

Reply:
Yes, once a case failed it is really hard to look out it. I'll use `TEST_P` to refactor this.